### DISCUSSION TITLE
Disallowing "r#" at the beginning of identifiers

### DISCUSSION THREAD

**Discussion by garrisonhh** [kdl-org/kdl#200] [Ideas]
As far as I can tell from the spec, `r#####` is a perfectly valid identifier. At the moment, given a sequence of characters representing a single kdl value or identifer, all types of values can be identified solely based upon their first two characters except for raw strings because of this particular case. For consistency, I'd also be in favor of disallowing `#` altogether from identifiers.

This is a tiny nitpick, but it makes sense to me from an implementer perspective!

**Comment by hkolbeck** [comment]
This definitely requires some attention. As much as it pains me, I'm in favor of disallowing `#` in bare identifiers. Currently kdl4j will reject bare identifiers starting with "r#". 

From an impl perspective, allowing such identifiers requires unbounded lookahead as far as I can figure, which is certainly prohibitive. Because kdl4j is hand written I can certainly tweak things to allow it, but others implementations may not be able.

**Comment by zkat** [comment]
I don't think it's a tiny nitpick. I think it's pretty significant, and warrants having a discussion with implementers about putting out a new spec.

**Comment by zkat** [comment]
Can someone volunteer to put up a PR that disallows `#`? And maybe we can start a discussion about disallowing a couple more symbols, because I need some for KQL? :)

**Reply by hkolbeck** [reply]
I can do that. I'd like to have the discussion about other symbols to be disallowed sooner rather than later if possible, so I can do them all at once.

**Reply by esoterra** [reply]
Is that a WIP KDL-based query language?
If so, you'll have to compete with [Kusto Query Language](https://docs.microsoft.com/en-us/azure/data-explorer/kql-quick-reference) for SEO

**Reply by hkolbeck** [reply]
@Kylebrown9 It is, the spec is here: https://github.com/kdl-org/kdl/blob/main/QUERY-SPEC.md

**Reply by esoterra** [reply]
Oh, I see. It's a way of querying KDL documents nodes. That sounds quite useful for some applications.
For some reason I was expecting it to be a way to express database queries in KDL.

**Comment by pingiun** [comment]
I believe this is also the case with numbers. Consider the token `+1214124` is this a number or an identifier? If we read the next character we can see it can only be an identifier `+12124123k` however if the next character were to be a space, this should've been read as a number instead. I think disallowing `+` and `-` at the start of identifiers would be a good fix for this.

**Reply by larsgw** [reply]
This was actually implemented already in #99 and then amended in #100 to allow for identifiers like `--foo`.

**Reply by pingiun** [reply]
Thanks for pointing me to this. I don't think it's explained very well in the prose and that's why I got confused. I'll take a look if I may be able to improve that wording

**Reply by larsgw** [reply]
Sounds good.
